Add raw editor toggle for using variables in flows operations

azrikahar opened the pull request:
## Description

Implements a MVP per-field toggle to make it possible to enter variables in any interfaces as discussed in https://github.com/directus/directus/pull/13801#issuecomment-1156565264.

https://user-images.githubusercontent.com/42867097/174791184-748111f6-6a60-4ee9-95be-bb1d30613f6c.mp4

Fixes #13787

## Type of Change

- [ ] Bugfix
- [x] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated

azrikahar committed:
add raw editor for flows operation options

azrikahar committed:
add comment to explain reasoning for watcher

azrikahar committed:
add simple raw editor with syntax highlighting

azrikahar commented:
Here's an attempt at adding minimal syntax highlighting:

![chrome_Z98VpD3Cqp](https://user-images.githubusercontent.com/42867097/174851900-5603f713-8978-4ac3-8002-cd60218ae10a.gif)

Technically I believe it may be possible to add autosuggestions as CodeMirror has an addon for it, but that definitely should fit in another future PR as that scope can creep quite abit (one example would be for the autocomplete to be _correct_, we'll need to make sure it only suggests all the operation keys _leading_ to it, and filter out others. This mean we'll need to pass it to the input as well)

>  but I wonder if there's a better icon for the "standard" input toggle. 

By "standard", do you mean the down arrow on the right side of the new raw editor toggle?

> Also, are all raw inputs single line? Would we want/need a multiline for others?

Yeah currently it's single line. Thought about this, but not sure what's a good approach to distinguish between single line or multiline, since they are all `string` field type. Unless in this particular scenario, it's okay to use the interface to make that determination? (which is fragile tbh). @rijkvanzanten thoughts on this?

benhaynes commented:
Amazing! Content editable?

> By "standard", do you mean the down arrow on the right side of the new raw editor toggle?

I mean the `code` icon we're using for the normal input (opposite the raw editor with the `data_object` icon). If the "standard" input is a dropdown, I don't think the code icon makes sense... maybe [something like `input`](https://fonts.google.com/icons?selected=Material+Icons+Outlined:dynamic_feed&icon.query=input)?

> Yeah currently it's single line.

Makes sense. Let's see what @rijkvanzanten thinks. I'm wondering about things like an email template, where line breaks might be useful/required... but perhaps the "standard" input will accept these objects in that case?

Also, instead of the current "text" placeholder (`Enter a value...`), should we put an example object (eg: `{{$trigger.xxx.yyy}}`) to be shorter and clearer?

azrikahar committed:
Merge branch 'main' into feat/raw-editor

azrikahar committed:
Add multiline to text fields & hide in json fields

azrikahar committed:
update input icon for toggle

azrikahar committed:
do not unset value for text fields

azrikahar commented:
> Amazing! Content editable?

It's using CodeMirror, which is the same used in our current Code interface. This is just a rendition of it as a whole new custom interface, lightly tweaked to support mustache syntax highlighting. Mainly the idea is technically it'll be "easier" to replace if we wish to use something else specifically for Raw Editor in the future.

> I mean the code icon we're using for the normal input (opposite the raw editor with the `data_object` icon). If the "standard" input is a dropdown, I don't think the code icon makes sense... maybe something like input?

Ahh makes perfect sense. I was fixated on the `code` icon acting as "there is an editor here which you can toggle", rather than the current state of the interface. 🤦‍♂️

> but perhaps the "standard" input will accept these objects in that case?

Yeah they do. I've now added multiline support for `text` fields in particular 👍 They were `string` as I mentioned, but using `text` actually makes sense both as a distinguishing factor and the WYSIWYG interface is _meant_ for text fields anyways.

I've also removed the toggle for json fields (such as query, filters etc) since it feels redundant to have the raw input for them. Do let me know if there's any perceivable drawback with that train of thought.

> Also, instead of the current "text" placeholder (`Enter a value...`), should we put an example object (eg: `{{$trigger.xxx.yyy}}`) to be shorter and clearer?

Agreed! Changed it to `{{$trigger.payload.example}}`.

---

With current changes:

https://user-images.githubusercontent.com/42867097/175024362-2fc5090e-642d-4b24-8aa2-3d5f9f05f0ad.mp4

azrikahar committed:
fix mustache tag value checking

azrikahar committed:
Merge branch 'main' into feat/raw-editor

azrikahar committed:
Merge branch 'main' into feat/raw-editor

azrikahar committed:
enable raw editor for Insights

azrikahar committed:
Merge branch 'main' into feat/raw-editor

azrikahar committed:
remove lint warning

azrikahar committed:
Merge branch 'main' into feat/raw-editor

rijkvanzanten committed:
Merge branch 'main' into feat/raw-editor

rijkvanzanten committed:
Reduce size + inline icons

rijkvanzanten commented on app/src/components/v-form/v-form.vue:
```diff
@@ -178,6 +186,24 @@ export default defineComponent({
 		const { toggleBatchField, batchActiveFields } = useBatch();
+		const { toggleRawField, rawActiveFields } = useRawEditor();
+
+		watch(
+			fields,
+			() => {
+				// make sure to toggle existing values with mustache tags as rawActiveFields
+				// to prevent the initial incompatible values from breaking the original interfaces
+				// on first load
+				for (const formField of formFields.value) {
+					const fieldValue = values.value?.[formField.field] ?? props.initialValues?.[formField.field];
+					if (!fieldValue || typeof fieldValue !== 'string') continue;
+					if (fieldValue.includes('{{') && fieldValue.includes('}}')) {
+						rawActiveFields.value.add(formField.field);
+					}
+				}
+			},
+			{ immediate: true }
+		);
```
Instead of auto-switching to raw, we should make sure that the individual interfaces don't crash on unexpected values. I think we can remove this block in favor of tweaking the other interfaces to show a notice if the value can't be edited (eg when `tags` doesn't get an array etc)

azrikahar replied:
It's been a hot minute, but I think another reason I did this was to also make it so interfaces such as tags/toggles (and technically somewhat relevant for dropdowns) to be "clearer" that they are using mustache values 🤔

Here's how the tags and toggle fields will look like initially when we edit an operation:

https://user-images.githubusercontent.com/42867097/179159248-32e73fcf-2a26-4b7c-963e-71721a4c55bb.mp4

The crashes should be fixed! This is more so a question of visual/UX I suppose. And I'm not 100% sure what is the intended notice should look like or how can we pivot that for this use case 😅

benhaynes commented:
I think the foreground color shift makes sense, and we also do the background color but at our light opacity/value (when active).

rijkvanzanten committed:
Merge branch 'main' into feat/raw-editor

rijkvanzanten reviewed (changes requested):
![CleanShot 2022-07-14 at 15 28 10@2x](https://user-images.githubusercontent.com/9141017/179067452-b35f1143-825b-4369-92ef-bea0ae494180.png)

~For some reason, code interfaces don't get the raw toggle? Is that on purpose @azrikahar?~


Nevermind, I should read the whole thread before commenting 🙈 

> I've also removed the toggle for json fields (such as query, filters etc) since it feels redundant to have the raw input for them. Do let me know if there's any perceivable drawback with that train of thought.

Lets keep it consistent and show it everywhere at all times 👍🏻 I was immediately confused as to why I couldn't switch to the raw input for JSON fields 🙂

rijkvanzanten commented on app/src/components/v-form/form-field-interface.vue:
```diff
@@ -30,6 +30,13 @@
+		<interface-system-raw-editor
+			v-else-if="rawEditorEnabled && rawEditorActive"
+			:value="modelValue === undefined ? field.schema?.default_value : modelValue"
+			:multi-line="field.type === 'text'"
```
Should the system interface just auto-switch to multi-line when `type === 'text'` is passed, rather than a new interface option?

```suggestion
			:type="field.type"
```

and then in the interface switch based on the `type` prop 🤔

azrikahar replied:
Good call. Definitely more flexible this way! Implemented 👍

rijkvanzanten commented:
@azrikahar Mind resolving the following? I think we can merge this before the end of the week 🙂 

- [ ] Add background primary-alt highlight to active toggle
- [ ] Remove watcher that auto-switches to raw mode for values containing `{{}}`
- [ ] Show raw toggle for all field types

❤️

azrikahar committed:
add background-highlight when active toggle

azrikahar committed:
change multiline prop to type prop

azrikahar committed:
show toggle for all field types (including json)

azrikahar committed:
remove watcher to toggle rawEditor on load

azrikahar committed:
fix raw editor emit

azrikahar committed:
fix request operation headers field type json

azrikahar committed:
fix raw editor value passed to codemirror

azrikahar committed:
prevent tags from crashing

azrikahar committed:
do not unset values anymore when toggling raw

azrikahar commented:
> Add background primary-alt highlight to active toggle

Here's how it looks like now. Hopefully that's the right direction, but feel free to tweak it!

![image](https://user-images.githubusercontent.com/42867097/179158852-f3db70ad-c088-4bcb-8361-9af3a8b4ad5b.png)

rijkvanzanten committed:
Merge branch 'main' into feat/raw-editor

rijkvanzanten commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -315,6 +315,8 @@ primary_key: Primary Key
 finish_setup: Finish Setup
 dismiss: Dismiss
+toggle_raw_editor: Toggle Raw Editor
+raw_editor_placeholder: "{'{'}{'{'}$trigger.payload.example{'}'}{'}'}"
```
It's a work of art really `"{'{'}{'{'} (╯°□°）╯︵ ┻━┻    {'}'}{'}'}"`
